**function with arguemnt**

In [2]:
def is_anagram(s1,s2):
    if sorted (s1) == sorted(s2):
        print('anagram')
    else:
        print('not anagram')

s1=input('Enter first string')
s2=input('Enter second string')

is_anagram(s1,s2)

Enter first string listen
Enter second string silent


anagram


In [10]:
def square_cube(n):
    print('square=',n**2)
    print('cube=',n**3)
n=eval(input('Enter a number'))
square_cube(n)

Enter a number 2


square= 4
cube= 8


In [3]:
def square_cube(n):
    square = n**2
    cube = n**3
    return square,cube
n= int(input(('Enter a number')))
square,cube =square_cube(n)
print('square=',square)
print('cube = ', cube)

Enter a number 2


square= 4
cube =  8


In [30]:
def intersection(list1,list2):
    s=[]
    for i in list1:
        if i in list2 and i not in s:
            s.append(i)
    return s

list1=eval(input('Enter list1 value '))
list2=eval(input('Enter list2 value'))

intersection(list1,list2)

Enter list1 value  [1,2,3,4,5]
Enter list2 value [3,4,5]


[3, 4, 5]

In [4]:
def vowels(str1):
    c=0
    s=[]
    for i in str1:
        if i in 'aeiou' and i not in s:
            print(i)
            s.append(i)
            c=c+1
    return s,c
str1=input('Enter str1')
vowels(str1)

Enter str1 aeroplane


a
e
o


(['a', 'e', 'o'], 3)

In [13]:
def two_sum(numbers,target):
    for i in range(len(numbers)):
        for j in range(i + 1,len(numbers)):
            if numbers[i] + numbers [j] == target :
                return i,j
numbers=[1,3,5,7,9]
target=10
two_sum(numbers,target)

(0, 4)

In [8]:
numbers = list(map(int, input("Enter numbers: ").split(',')))

print(numbers)
print(len(numbers))

Enter numbers:  1,3,5,7


[1, 3, 5, 7]
4


**default argumrnt**

In [10]:
import random
import string 

def generate_password(length=8):
    characters = string.ascii_letters + string.digits+ string.punctuation
    password=''

    for i in range(length):
        password = password + random.choice(characters)
        
    return password
generate_password()


'p80T;pr8'

In [37]:
generate_password(12)  
    

'GM;[dTW|:?$i'

In [9]:
import string
dir(string)

['Formatter',
 'Template',
 '_ChainMap',
 '__all__',
 '__builtins__',
 '__cached__',
 '__doc__',
 '__file__',
 '__loader__',
 '__name__',
 '__package__',
 '__spec__',
 '_re',
 '_sentinel_dict',
 '_string',
 'ascii_letters',
 'ascii_lowercase',
 'ascii_uppercase',
 'capwords',
 'digits',
 'hexdigits',
 'octdigits',
 'printable',
 'punctuation',
 'whitespace']

In [ ]:
dir(str)
   ↓
String methods/tools

dir(string)
   ↓
Tools/constants inside the string module

In [1]:
def simple_interest(p,t,r=5):
    
    interest=(p*t*r / 100)
    return interest

simple_interest(10000,1)

500.0

In [45]:
simple_interest(20000,3,4)

2400.0

In [56]:
def table(number=5):
    for i in range(1,11):
        print(number ,'x' ,i , '=', number * i)

table()
    

5 x 1 = 5
5 x 2 = 10
5 x 3 = 15
5 x 4 = 20
5 x 5 = 25
5 x 6 = 30
5 x 7 = 35
5 x 8 = 40
5 x 9 = 45
5 x 10 = 50


In [58]:
table(6)

6 x 1 = 6
6 x 2 = 12
6 x 3 = 18
6 x 4 = 24
6 x 5 = 30
6 x 6 = 36
6 x 7 = 42
6 x 8 = 48
6 x 9 = 54
6 x 10 = 60


In [60]:
def area_of_rectangle(length,width=10):
    area=length * width 
    return area

area_of_rectangle(2)

20

In [61]:
area_of_rectangle(4,2)

8

In [6]:
def validate_password(password,min_length=8):
    if len(password) < min_length:
        return 'invalid password'
    
    has_upper=False
    has_lower=False
    has_digit=False
    has_speacial=False

    speacial='!@#$%^&*'

    for ch in password:
        if ch.isupper():
            has_upper=True
        elif ch.islower():
            has_lower=True
        elif ch.isdigit():
            has_digit=True
        elif  ch in speacial:
            has_speacial=True

    if has_upper and has_lower and has_digit and has_speacial:
        return 'valid password'
    else:
        return 'invalid password'

validate_password('Abc@1234')
validate_password('Sar@12',6)

'valid password'

In [5]:
def validate_password(password):
    
    has_upper=False
    has_lower=False
    has_digit=False
    has_speacial=False

    speacial='!@#$%^&*'

    for ch in password:
        if ch.isupper():
            has_upper=True
        elif ch.islower():
            has_lower=True
        elif ch.isdigit():
            has_digit=True
        elif  ch in speacial:
            has_speacial=True

    if has_upper and has_lower and has_digit and has_speacial:
        return 'valid password'
    else:
        return 'invalid password'

validate_password('Abc@1')

'valid password'

In [81]:
validate_password('sarvesh')

'invalid password'

'valid password'

**recursion (factorial type problem)**

In [110]:
def fact(n):
    if n ==0 or n==1:
        return 1
    else:
        return n * fact(n-1)
n=eval(input('Enter a number '))     
print('factorial of number is',fact(n))  

Enter a number  3


factorial of number is 6


In [12]:
def reverse(text):
    if text == '':
        return ''
    else:
        return text[-1] + reverse(text[0:-1])
reverse('romeo')

'oemor'

In [ ]:
hello
 ↓
ello + h
 ↓
llo + e + h
 ↓
lo + l + e + h
 ↓
o + l + l + e + h
 ↓
olleh

In [13]:
def reverse(text):
    if text == '':
        return ''
    else:
        return reverse(text[1:]) + text[0]   # ✅ CORRECT

print(reverse('richa'))

ahcir


In [130]:
def fibonacci(n):
    if n==0:
        return 0
    elif n==1:
        return 1
    else:
        return fibonacci(n-1) + fibonacci(n-2)
fibonacci(10)

55

In [ ]:
                 fibo(5)
                /       \
           fibo(4)      fibo(3)
           /    \       /    \
       fibo(3) fibo(2) fibo(2) fibo(1)
       /   \    /  \    /  \
    fibo(2) 1  1    0  1    0
     /  \
    1    0

In [8]:
def decimal_to_binary(n):
    if n <=1:
        print(n,end=' ')
        return 

    decimal_to_binary(n //2)
    print(n % 2, end=' ')
n=eval(input('Enter number '))    
print('binary=',end=' ')
decimal_to_binary(n)

Enter number  4


binary= 1 0 0 

In [175]:
def palindrome(word):
    if len(word)<=1:
        return True 
        
    if word[0] != word[-1]:
        return False 
    
    return palindrome(word[1:-1])


word=input('Enter a word')
if palindrome(word):
    print('palindrome ')
else:
    print('Not palindrome ')

Enter a word madam


palindrome 


**args** and **kwargs**

In [184]:
def add(*args):
    sum =0
    for i in args:
        sum = sum + i
    return sum

add(10,20)

30

In [183]:
add(1,2,3,4,5)

15

In [209]:
def prime(*args):
    result=[]
    for i in args:
        if i > 1:
            for j in range(2,i):
                if i % j==0:
                    break 
            else:
                result.append(i)
    return result

prime(1,2,3,5,7,8)
    

[2, 3, 5, 7]

In [20]:
#3. Write a function that accepts **kwargs and prints each key-value pair.
def student(**kwargs):
    for key,value in kwargs.items():
        print(key,'=',value)
student(name='sarvesh',rollno =21,branch = 'csds')

name = sarvesh
rollno = 21
branch = csds


In [13]:
def student(*args,**kwargs):
    
    total=sum(args)
    avg=total / len(args)

    for key,value in kwargs.items():
        print(key,'=',value)

    print('Total=',total)
    print('avg=',avg) 
    
    if avg >= 90:
        print('A')
    elif avg >=80:
        print('B')
    elif avg >= 70:
        print('C')
    elif avg >= 60:
        print('D')
    elif avg >= 50:
        print('E')
    else:
        print('fail')
student(90,80,85,name='richa',rollno=21,branch='cse')

name = richa
rollno = 21
branch = cse
Total= 255
avg= 85.0
B


In [16]:
def student1(**kwargs):
    for key,value in kwargs.items():
        print(key,'=',value)
student1(name='sarbe',age='20',courese='data science')

name = sarbe
age = 20
courese = data science


**return statement**

In [20]:
def square(n):
    return n*n
n=eval(input('Enter number '))
print('square of number =',square(n))

Enter number  4


square of number = 16


In [23]:
def longest(word):
    return word.split()
word=input('Enter word :')
longest(word)

Enter word : i love datascience for treading 


['i', 'love', 'datascience', 'for', 'treading']

In [24]:
len(word)

32

In [8]:
def divisiable(num):
    if num % 3 == 0 and num % 5==0:
        return True
    else:
       return  False
num=eval(input('Enter a  number :'))
divisiable(num)

Enter a  number : 15


True

In [35]:
def divisiable(start,end):
    num=[]
    for i in range(start,end+1):
        if i % 3==0 and i % 5==0:
            num.append(i)
            
    return num 

start=eval(input('Enter the starting num :'))
end=eval(input('Enter the  ending number :.'))

divisiable(start,end)



Enter the starting num : 1
Enter the  ending number :. 50


[15, 30, 45]

In [46]:
def count_character(text):
    uppercase=0
    lowercase=0
    digit=0

    for i in text:
        if i.isupper():
            uppercase = uppercase + 1
        elif i.islower():
            lowercase = lowercase +1
        elif i.isdigit():
            digit = digit + 1
        else:
            pass

    return uppercase,lowercase,digit

text=input('Enter a string :')
result = count_character(text)

            
print('uppercase=',result[0])
print('lowercase=',result[1])
print('digits=',result[2])

Enter a string : fefwfqd1SS


uppercase= 2
lowercase= 7
digits= 1


In [54]:
def amstrong(num):
    original = num   
    power=len(str(num))
    total = 0

    while num  >0 :
        digit = num %  10
        total = total + (digit ** power) 
        num = num // 10

    if total == original :
       return True 
    else:
       return False
       
num=eval(input('Enter a number :'))

amstrong(num)

Enter a number : 153


True

| Step                  | Calculation |  Result |
| --------------------- | ----------- | ------: |
| `digit = num % 10`    | `153 % 10`  |   **3** |
| `digit ** power`      | `3 ** 3`    |  **27** |
| `total = total + 27`  | `0 + 27`    |  **27** |
| `num = num // 10`     | `153 // 10` |  **15** |
| **Variable:** `num`   |             |  **15** |
| **Variable:** `digit` |             |   **3** |
| **Variable:** `total` |             |  **27** |
| `digit = num % 10`    | `15 % 10`   |   **5** |
| `digit ** power`      | `5 ** 3`    | **125** |
| `total = total + 125` | `27 + 125`  | **152** |
| `num = num // 10`     | `15 // 10`  |   **1** |
| **Variable:** `num`   |             |   **1** |
| **Variable:** `digit` |             |   **5** |
| **Variable:** `total` |             | **152** |
| `digit = num % 10`    | `1 % 10`    |   **1** |
| `digit ** power`      | `1 ** 3`    |   **1** |
| `total = total + 1`   | `152 + 1`   | **153** |
| `num = num // 10`     | `1 // 10`   |   **0** |
| **Variable:** `num`   |             |   **0** |
| **Variable:** `digit` |             |   **1** |
| **Variable:** `total` |             | **153** |


**global vs local**

In [61]:
x=100
def show():
    y =200
    c= x+10
    print('c=',c)
    print('Global variable=',x)
    print('local variable=',y)

show()

print('Global variable =',x)

c= 110
Global variable= 100
local variable= 200
Global variable = 100


In [13]:
var =20
def modify():
    global var
    var = var+ 100
    print('inside fun ,global var =',var)

print('before function call = ',var)

modify()

print('after function call =',var )


before function call =  20
inside fun ,global var = 120
after function call = 120


In [11]:
a = 100
def observe():
    a =200
    print('inside function a =',a)
    
print('before function call a =',a)

observe()

print('after function call a = ',a)

before function call a = 100
inside function a = 200
after function call a =  100


In [9]:
counter = 0
def fun():
    global counter
    counter += 1
    print('function called' ,counter,'times')
    
fun()
fun()
fun()
print('final counter=',counter)

function called 1 times
function called 2 times
function called 3 times
final counter= 3


In [87]:
var = 100
def fun():                                     # x is accessible only inside the function.
    x=200
    print('local x=',x)
    print('global var=',var)
print('gloabl var =',var)
fun()

gloabl var = 100
local x= 200
global var= 100


- var is accessible everywhere.
- x is accessible only inside the function.